### PR TITLE
[TTNN] Avoid degenerate tile padding in reshape -> to_tensor_spec

### LINKED ISSUE #8887: `scatter` operand reshape to 1-D emits TILE layout - degenerate height 1→32 padding causes 32× DRAM blow-up (OOMs GNN training), even though the next op converts it to row-major
### Describe the bug

Follow-up to #8714 / #8718. With that fix, stock `GATv2Conv` scatter now **compiles** and runs forward/inference natively on TT - thank you. But full-graph **training** on PubMed now hits a DRAM OOM that traces to a **lowering inefficiency**, not GAT math.

When a scatter source operand of shape `[E, F]` is flattened to 1-D `[E*F]` before the scatter, the reshape produces the 1-D tensor in **TILE** layout. A 1-D tensor is mapped to `[1, E*F]` and tiled into 32×32 tiles, so the **degenerate height 1 pads up to a full tile of 32 rows → a 32× memory blow-up**. The very next op converts that buffer to **row-major** (which is what `ttnn.scatter` actually consumes), so the tile-padded buffer is **pure transient waste** - but its allocation alone overflows DRAM.

For GATv2 on full PubMed the wasted transient is **887 MB** for a logical tensor of only ~28 MB. The op is correct; it just allocates 32× more than needed.

**IR from the compiled training graph** (message-aggregation scatter operand prep):

```mlir
// %120 : per-edge message [E_sl, heads*out]
%121 = "ttnn.reshape"(%120) <{shape = [6935360 : i32]}>
     : (tensor<108365x64xf32, #ttnn_layout46>)
    -> tensor<6935360xf32, #ttnn_layout57>          // 1-D, TILE layout  ← the blow-up
%128 = "ttnn.to_layout"(%121) <{layout = #ttnn.layout<row_major>}>
     : (tensor<6935360xf32, #ttnn_layout57>)
    -> tensor<6935360xf32, #ttnn_layout60>          // converted to ROW-MAJOR (~28 MB)
%129 = "ttnn.scatter"(%5, %127, %128) <{dim = 0, scatter_reduce_type = #ttcore.reduce_type<sum>}>
     : (tensor<1261888xf32>, tensor<6935360xsi32>, tensor<6935360xf32>) -> tensor<1261888xf32>
```

The TILE-layout 1-D intermediate vs. the row-major one the scatter consumes:

```mlir
#ttnn_layout57 = …memref<1x216730x!ttcore.tile<32x32, f32>, #dram>…   // %121: TILE  → 887 MB
#ttnn_layout60 = …memref<1x6935360xf32, #dram>…                       // %128: ROW-MAJOR → 28 MB
```

**Byte math (the 32×):**

| Tensor | shape / layout | bytes |
|---|---|---|
| logical data | `6,935,360 × f32` | `27,741,440` (~28 MB) |
| `%121` TILE `[1, 6935360]` | `216,730 tiles × 32×32 × 4 B` | **`887,726,080`** (~847 MiB) |

`887,726,080 / 27,741,440 = 32.0` exactly — the single real row (height 1) padded to a full 32-row tile. `887,726,080 B` is the exact size of the failing DRAM allocation:

```
TT_FATAL: Out of Memory: Not enough space to allocate 887726080 B DRAM buffer across 12 banks …
(allocated: 969296416 B, free: 101476768 B, largest free block: 55486464 B)
```

(The sibling softmax-denominator scatter flattens `108365 × 8 = 866,920` → `1×27092` tiles ≈ 110 MB, which fits; the `×64` message scatter is the one that overflows.)

### Steps to reproduce the issue

Minimal, PyG-free reproducer of the `reshape → to_layout(row_major) → ttnn.scatter` operand sequence at PubMed dimensions:

```python
import os
os.environ.setdefault("TTXLA_LOGGER_LEVEL", "DEBUG")
os.environ["XLA_STABLEHLO_COMPILE"] = "1"
import torch, torch_xla
import torch_xla.runtime as xr
xr.set_device_type("TT")

torch_xla.set_custom_compile_options({"export_path": "./irs_repro", "export_tensors": True})

E, F, N = 108365, 64, 19717          # PubMed edges+self-loops, heads*out, nodes
dev = torch_xla.device()
index = torch.randint(0, N, (E,), dtype=torch.long, device=dev)
src   = torch.randn(E, F, device=dev)
out   = torch.zeros(N, F, device=dev)
out = out.scatter_add(0, index.unsqueeze(-1).expand(-1, F), src)
torch_xla.sync(wait=True)
```

1. Run the script on a single N300 with a build containing the #8718 fix.
2. Open the exported `./irs_repro/.../ttnn_*.mlir`.
3. Find the `ttnn.reshape` to `[E*F]` that feeds `ttnn.scatter`.
4. Note its layout is `memref<1x…x!ttcore.tile<32x32>>` (TILE) immediately before a `ttnn.to_layout<row_major>` — i.e. a tile-padded 1-D buffer that is discarded for a row-major copy. At PubMed scale this transient is 887 MB and OOMs DRAM in the backward pass of full GATv2 training.

(Originally observed end-to-end: stock `GATv2Conv` (`use_spmm=false`) training on full PubMed — forward/eval runs on device, the first training step OOMs with the `887726080 B` allocation above.)

### Expected behavior

The flattened scatter operand should not be materialized in tile layout with a padded degenerate dim. Either:

- emit the `reshape` result directly in **row-major** (the consumer `to_layout` converts to row-major anyway — fuse/fold it), or
- scatter on the `[E, F]` shape without flattening to a degenerate `[1, E*F]` tile tensor.

Either removes the 32× transient (887 MB → ~28 MB) and lets stock `GATv2Conv` **train** on full PubMed on a single N300.

### Please complete the following environment information

- `pjrt-plugin-tt==1.3.0.dev20260624002726` (tt-mlir `70ff200c…`, tt-metal `7f6364a1`, built 2026-06-23) — contains the #8718 fix
- `torch_xla 2.9.0+git8d31cb3`
- Wormhole b0 (N300), single chip (`mesh_shape: null`), KMD 2.7.0, fw bundle 19.5.0

**Related:** #8714 / #8718 (scatter compile blow-up - fixed; this is the next wall it exposed); #8321 (`grid_sample` tile-padded-constant DRAM OOM, 32× blow-up - sibling *symptom*, different trigger op).

### PR DESCRIPTION
**Pull request by igrn256**
### Ticket

Closes #8887

### Problem description

In the scatter lowering the `[E, F]` operand is flattened to 1-D `[E*F]`, and `ttnn.reshape` produces that result in `TILE` layout. A 1-D shape is laid out as `[1, E*F]`, so with 32x32 tiles the degenerate height of 1 is padded out to 32 rows. The very next op converts the buffer back to row-major - that is what `ttnn.scatter` actually consumes - so the tiled buffer is pure transient waste.

For the GATv2 / full-PubMed training case in the issue that is 887,726,080 B (216,730 tiles x 4,096 B) where 27,741,440 B (6,935,360 elements x 4 B) is needed: exactly 32.00x, and enough to push an N300 into DRAM OOM. (The comment on the issue quoted these in mixed units; the byte counts here are the exact ones.)

There is merged downstream waiting on this: tt-blacksmith [#611](https://github.com/tenstorrent/tt-blacksmith/pull/611) routed around the bug via SpMM and explicitly left the stock `use_spmm: false` path in the tree pending a fix for #8887.

### What's changed

New pattern `ToTensorSpecRowMajorAdjusting<OpTy>` in `TTNNMemoryManagement.cpp`, instantiated for `ttnn::ReshapeOp`. It matches `op -> to_tensor_spec(row_major)` where `op` has a single use, converts the input to row-major up front, and rrebuilds `op` by cloning it with the consumer's encoding, so the tiled 1-D result is never materialized and no trailing conversion is needed.

As suggested by @djstefanovicTT in the issue, the pattern is templated on `OpTy` and instantiated only for `ReshapeOp` for now, so the scope stays narrow. Rebuilding through `rewriter.clone` rather than `create<ttnn::ReshapeOp>` is what keeps it op-agnostic, since `getShapeAttr()` is `ReshapeOp`-specific.

The rewrite fires only when `getShardSizeInBytes()` for the tiled layout exceeds the row-major one. There is deliberately no size threshold, unlike the neighbouring `PermuteRowMajorAdjusting` / `ReshapeRowMajorAdjusting`: those *add* a layout conversion and so have to amortize its cost, whereas here the conversion already exists in the incoming IR and is only moved earlier in the chain - there is nothing to amortize.

Whether the input needs converting is decided from `inputLayout.isTiled()` rather than from the kind of op that produced it, since a `to_tensor_spec` producer can hand back `TILE`.

#### Changes in tests

Six functions in `test/ttmlir/Transforms/memory_management.mlir`:

- `reshape_to_tensor_spec_row_major` - the rewrite itself.
- `reshape_to_tensor_spec_tile_aligned` - negative. Differs from the positive case *only* in tile alignment, so the pattern not firing is attributable to the footprint guard and nothing else.
- `reshape_to_tensor_spec_l1_consumer` - consumer in L1 rather than DRAM, covered by the same single conversion.
- `reshape_to_tensor_spec_multi_use` - negative, locking the `hasOneUse` boundary.
- `reshape_to_tensor_spec_tiled_input` - input produced by a `to_tensor_spec` that yields `TILE`, covering the layout-based gate.
- `reshape_to_tensor_spec_unaligned_flatten` - flattened length not a multiple of 32.

Not verified on device: I don't have TT hardware. `check-ttmlir` is green locally and `git clang-format` is clean; the two `StableHLO/ComplexDataTypeConversion` failures reproduce on an unmodified `origin/main` and are unrelated to this diff. Silicon coverage would have to come from CI.

### Checklist

- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: test/ttmlir/Transforms/memory_management.mlir
Comment (djstefanovicTT) [inline test/ttmlir/Transforms/memory_management.mlir:1]: Too many verbose comments that do not add any real value. Take a look at how other patterns are structured. It is enough for only a single short comment expressing which pattern is targeted and all relevant checks. For example:
```
  // a 1-D reshape whose only consumer is a to_tensor_spec(row_major) must emit its result directly in row-major
  // CHECK: %[[RM_IN:.*]] = "ttnn.to_tensor_spec"(%arg0)
  // CHECK: "ttnn.reshape"(%[[RM_IN]])
  // CHECK-SAME: shape = [6935360 : i32]
  // CHECK-SAME: -> tensor<6935360xf32
  // CHECK-NOT: "ttnn.to_tensor_spec"
  // CHECK: return
```

File: lib/Dialect/TTNN/Transforms/TTNNMemoryManagement.cpp [hunk 1]
```diff
@@ -1142,6 +1142,139 @@ class ReshapeRowMajorAdjusting : public OpRewritePattern<ttnn::ReshapeOp> {
...
+// instantiated for ttnn::ReshapeOp only.
+//
+// Example, for OpTy = ttnn::ReshapeOp:
+//   reshape [108365, 64] TILE -> [6935360] TILE
+//   to_tensor_spec [6935360] TILE -> [6935360] ROW_MAJOR
+//
+// becomes:
+//
+//   to_tensor_spec [108365, 64] TILE -> [108365, 64] ROW_MAJOR
+//   reshape [108365, 64] ROW_MAJOR -> [6935360] ROW_MAJOR
```
Comment (djstefanovicTT) [inline lib/Dialect/TTNN/Transforms/TTNNMemoryManagement.cpp:1163]: It is not recommended to mention resolved issues. Also comments should be more concise to avoid bloat.

```suggestion
// This pattern pulls the row-major conversion above an op whose only consumer
// is a to_tensor_spec that converts that op's result to row-major.
//
// An op that produces a non tile-aligned result in TILE layout pads its
// degenerate height to a full tile, inflating the DRAM footprint. When the next
// op immediately converts that result to row-major, the padded TILE buffer is
// pure transient waste.
//
// The rewrite converts the input to row-major first and re-emits the op
// directly in row-major, avoiding materialization of the padded TILE buffer.
//
// Example, for OpTy = ttnn::ReshapeOp:
//   reshape [108365, 64] TILE -> [6935360] TILE
//   to_tensor_spec [6935360] TILE -> [6935360] ROW_MAJOR
//
// becomes:
//
//   to_tensor_spec [108365, 64] TILE -> [108365, 64] ROW_MAJOR
//   reshape [108365, 64] ROW_MAJOR -> [6935360] ROW_MAJOR
```

File: lib/Dialect/TTNN/Transforms/TTNNMemoryManagement.cpp [hunk 2]
```diff
@@ -1142,6 +1142,139 @@ class ReshapeRowMajorAdjusting : public OpRewritePattern<ttnn::ReshapeOp> {
...
+    // Skip unless tiling actually inflates the memory footprint. Compare
+    // physical shard sizes rather than tensor shapes: for a tile-padded 1-D
+    // result, the 32x inflation is reflected in the layout, not in the logical
+    // shape (see #8887).
+    //
+    // Unlike the sibling patterns, no fixed size threshold is used here. Those
+    // patterns introduce a new row-major conversion, so its cost must be
+    // amortized by a sufficiently large memory saving. Here the row-major
+    // conversion already exists in the input IR and is simply moved before the
+    // op, so there is no additional conversion cost to amortize.
```
Comment (djstefanovicTT) [inline lib/Dialect/TTNN/Transforms/TTNNMemoryManagement.cpp]: +1 for excess comments
```suggestion
    // Skip unless tiling actually inflates the memory footprint. 
```

File: lib/Dialect/TTNN/Transforms/TTNNMemoryManagement.cpp [hunk 3]
```diff
@@ -1142,6 +1142,139 @@ class ReshapeRowMajorAdjusting : public OpRewritePattern<ttnn::ReshapeOp> {
...
+    auto toTensorSpecResultType =
+        mlir::cast<RankedTensorType>(toTensorSpecUser.getResult().getType());
+    auto toTensorSpecResultLayout = mlir::dyn_cast<ttnn::TTNNLayoutAttr>(
+        toTensorSpecResultType.getEncoding());
+    if (!toTensorSpecResultLayout || toTensorSpecResultLayout.isTiled()) {
+      return failure();
+    }
+
+    auto rowMajorOpLayout =
+        TTNNLayoutAttr::Builder(opResultLayout, opResultType.getShape())
```
Comment (djstefanovicTT) [inline lib/Dialect/TTNN/Transforms/TTNNMemoryManagement.cpp]: Just `TTNNLayoutAttr::Builder(opResultType)` is enough.

File: lib/Dialect/TTNN/Transforms/TTNNMemoryManagement.cpp [hunk 4]
```diff
@@ -1142,6 +1142,139 @@ class ReshapeRowMajorAdjusting : public OpRewritePattern<ttnn::ReshapeOp> {
...
+    Value replacement = newOp->getResult(0);
+    if (rowMajorOpResultType != toTensorSpecResultType) {
+      auto reconciled = utils::createToTensorSpecOp(
+          toTensorSpecUser,
+          mlir::cast<mlir::TypedValue<RankedTensorType>>(newOp->getResult(0)),
+          rewriter, toTensorSpecResultLayout.getLayout(),
+          toTensorSpecResultLayout.getBufferType(),
+          toTensorSpecResultLayout.getMemLayout(),
+          toTensorSpecResultLayout.getDataType(), "_reconcile");
+      replacement = reconciled.getResult();
```
Comment (djstefanovicTT) [inline lib/Dialect/TTNN/Transforms/TTNNMemoryManagement.cpp]: It should be possible that this pattern can resolve all encoding with only one to_tensor_spec op, without trailing reconciliatory change.

**Review by djstefanovicTT — COMMENTED** [review]
Thanks for the work! There are some suggestions, please review them and update the code before CI.

File: test/ttmlir/Transforms/memory_management.mlir
Comment (igrn256) [reply test/ttmlir/Transforms/memory_management.mlir:1]: Trimmed to a single line per case.

File: lib/Dialect/TTNN/Transforms/TTNNMemoryManagement.cpp (hunk 1 above)
Comment (igrn256) [reply lib/Dialect/TTNN/Transforms/TTNNMemoryManagement.cpp:1163]: Applied.

File: lib/Dialect/TTNN/Transforms/TTNNMemoryManagement.cpp (hunk 2 above)
Comment (igrn256) [reply lib/Dialect/TTNN/Transforms/TTNNMemoryManagement.cpp]: Applied.

File: lib/Dialect/TTNN/Transforms/TTNNMemoryManagement.cpp (hunk 3 above)
Comment (igrn256) [reply lib/Dialect/TTNN/Transforms/TTNNMemoryManagement.cpp]: Applied.

File: lib/Dialect/TTNN/Transforms/TTNNMemoryManagement.cpp (hunk 4 above)
Comment (igrn256) [reply lib/Dialect/TTNN/Transforms/TTNNMemoryManagement.cpp]: Done. The reshape takes the consumer's encoding directly now; `_reconcile` is now `_l1_consumer`.
Two related changes while I was in there: dropped the `rowMajorOpLayout == opResultLayout` check (unreachable after the `isTiled()` guard above), and added a `dtype` early exit, since without the reconciling op the reshape would inherit a dtype` it can't express. Happy to split either out if you'd rather keep this minimal.

**Comment by igrn256** [comment]
Addressed all five, thanks.

**Review by djstefanovicTT — APPROVED** [review]
Thank you! Looks much better now.

**Comment by igrn256** [comment]
Thanks for the review — the reconciliation catch made this noticeably cleaner.
